### PR TITLE
Bugfix `feature class` with threads

### LINKED ISSUE #20803: BBC: Blead Breaks IO::Async
Passes on 5.37.8.

On blead:

```
$ bleadperl -Mblib t/41routine.t 
ok 1 - '$routine for fork model' isa 'IO::Async::Routine'
ok 2 - $routine has refcount 1 initially for fork model
ok 3 - $routine has refcount 2 after $loop->add for fork model
ok 4 - $routine->model for fork model
ok 5 - Result for fork model
ok 6 - $routine has refcount 2 before $loop->remove
ok 7 - $routine has refcount 1 before EOF
ok 8 - on_return for fork model
ok 9 - Result from ->result_future of return
ok 10 - on_die for fork model
ok 11 - Failure from ->result_future of die
ok 12 - Recv on closed channel for fork model
ok 13 - Routine is ready for SIGINT
ok 14 - Routine caught SIGINT
ok 15 - '$routine for thread model' isa 'IO::Async::Routine'
ok 16 - $routine has refcount 1 initially for thread model
Segmentation fault
```

Appears to be a stack blowup:

```
$ bleadperl --gdb -Mblib t/41routine.t 
...
(gdb) run
Starting program: /home/leo/perl5/perlbrew/perls/bleadperl/bin/perl -Mblib t/41routine.t
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
ok 1 - '$routine for fork model' isa 'IO::Async::Routine'
ok 2 - $routine has refcount 1 initially for fork model
[Detaching after fork from child process 2491048]
ok 3 - $routine has refcount 2 after $loop->add for fork model
ok 4 - $routine->model for fork model
ok 5 - Result for fork model
ok 6 - $routine has refcount 2 before $loop->remove
ok 7 - $routine has refcount 1 before EOF
[Detaching after fork from child process 2491049]
ok 8 - on_return for fork model
ok 9 - Result from ->result_future of return
[Detaching after fork from child process 2491050]
ok 10 - on_die for fork model
ok 11 - Failure from ->result_future of die
[Detaching after fork from child process 2491051]
ok 12 - Recv on closed channel for fork model
[Detaching after fork from child process 2491052]
ok 13 - Routine is ready for SIGINT
ok 14 - Routine caught SIGINT
ok 15 - '$routine for thread model' isa 'IO::Async::Routine'
ok 16 - $routine has refcount 1 initially for thread model

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7d39895 in __GI___libc_malloc (bytes=40) at ./malloc/malloc.c:3315
3315    ./malloc/malloc.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7d39895 in __GI___libc_malloc (bytes=40) at ./malloc/malloc.c:3315
#1  0x00005555557a291a in Perl_safesysmalloc (size=40) at util.c:161
#2  0x0000555555833784 in Perl_sv_grow_fresh (my_perl=0x555556cbdb20, sv=0x5555573f5478, newlen=16) at sv.c:1421
#3  0x0000555555858217 in Perl_sv_setpvn_fresh (my_perl=0x555556cbdb20, sv=0x5555573f5478, ptr=0x555555bc99f4 "", len=0) at sv.c:5029
#4  0x000055555587564e in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9551
#5  0x00005555557a53c8 in S_mess_alloc (my_perl=0x555556cbdb20) at util.c:1383
#6  0x00005555557a67a5 in Perl_vmess (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff280) at util.c:1673
#7  0x00005555557a7d5b in Perl_vcroak (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff280) at util.c:1892
#8  0x00005555557a80cf in Perl_croak_nocontext (pat=0x555555bff11a "%s") at util.c:1934
#9  0x00005555558d52dc in Perl_croak_memory_wrap () at perlstatic.h:24
#10 0x00005555558d6fb2 in Perl_tmps_grow_p (my_perl=0x555556cbdb20, ix=-6076574518398426621) at scope.c:218
#11 0x00005555558756b7 in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9564
#12 0x00005555557a53c8 in S_mess_alloc (my_perl=0x555556cbdb20) at util.c:1383
#13 0x00005555557a67a5 in Perl_vmess (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff4d0) at util.c:1673
#14 0x00005555557a7d5b in Perl_vcroak (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff4d0) at util.c:1892
#15 0x00005555557a80cf in Perl_croak_nocontext (pat=0x555555bff11a "%s") at util.c:1934
#16 0x00005555558d52dc in Perl_croak_memory_wrap () at perlstatic.h:24
#17 0x00005555558d6fb2 in Perl_tmps_grow_p (my_perl=0x555556cbdb20, ix=-6076574518398426622) at scope.c:218
#18 0x00005555558756b7 in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9564
#19 0x00005555557a53c8 in S_mess_alloc (my_perl=0x555556cbdb20) at util.c:1383
#20 0x00005555557a67a5 in Perl_vmess (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff720) at util.c:1673
#21 0x00005555557a7d5b in Perl_vcroak (my_perl=0x555556cbdb20, pat=0x555555bff11a "%s", args=0x7fffff7ff720) at util.c:1892
#22 0x00005555557a80cf in Perl_croak_nocontext (pat=0x555555bff11a "%s") at util.c:1934
#23 0x00005555558d52dc in Perl_croak_memory_wrap () at perlstatic.h:24
#24 0x00005555558d6fb2 in Perl_tmps_grow_p (my_perl=0x555556cbdb20, ix=-6076574518398426623) at scope.c:218
#25 0x00005555558756b7 in Perl_newSVpvn_flags (my_perl=0x555556cbdb20, s=0x555555bc99f4 "", len=0, flags=524288) at sv.c:9564
...
(continues for a long time)
```

In particular it appears to be stuck in infinite recursion trying to croak.

### PR DESCRIPTION
**Pull request by leonerd**
Fixes #20803

### CODE REVIEW

File: sv.c [hunk 1]
```diff
@@ -14746,70 +14810,8 @@ S_sv_dup_common(pTHX_ const SV *const ssv, CLONE_PARAMS *const param)
...
-                        daux->xhv_mro_meta = saux->xhv_mro_meta
-                            ? mro_meta_dup(saux->xhv_mro_meta, param)
-                            : 0;
-
-                        /* Record stashes for possible cloning in Perl_clone(). */
-                        if (HvNAME(ssv))
-                            av_push(param->stashes, dsv);
-                    }
+                    if (HvHasAUX(ssv))
+                        S_sv_dup_hvaux(aTHX_ ssv, dsv, param);
```
Comment (Leont) [inline sv.c]: This really should be called as `sv_dup_hvaux`

File: pad.c [hunk 2]
```diff
@@ -2478,6 +2478,7 @@ Perl_pad_push(pTHX_ PADLIST *padlist, int depth)
 #if defined(USE_ITHREADS)
 
 #  define av_dup_inc(s,t)	MUTABLE_AV(sv_dup_inc((const SV *)s,t))
+#  define hv_dup_inc(s,t)	MUTABLE_HV(sv_dup_inc((const SV *)s,t))
```
Comment (Leont) [inline pad.c]: These macro are also defined in `regcomp.c` and `sv.c`, maybe they should be moved to a header.

File: sv.c (hunk 1 above)
Comment (leonerd) [reply sv.c]: Where would I put the macro for that then? Do I have to do the full dance with embed.fnc and #ifdef-hiding it?

File: pad.c (hunk 2 above)
Comment (leonerd) [reply pad.c]: That sounds a good idea. Perhaps I'll tidy that up first as a separate prereq PR

File: sv.c (hunk 1 above)
Comment (wolfsage) [reply sv.c]: I think S_sv_dup_inc_multiple is good to copy from maybe. In embed.h there's already this section:

```
#     if defined(USE_ITHREADS)
#       define sv_dup_common(a,b)               S_sv_dup_common(aTHX_ a,b)
#       define sv_dup_inc_multiple(a,b,c,d)     S_sv_dup_inc_multiple(aTHX_ a,b,c,d)
#       define unreferenced_to_tmp_stack(a)     S_unreferenced_to_tmp_stack(aTHX_ a)
#     endif /* defined(USE_ITHREADS) */
```

File: pad.c (hunk 2 above)
Comment (leonerd) [reply pad.c]: See https://github.com/Perl/perl5/pull/20808

**Review by Leont — APPROVED** [review]

**Pull request merged**